### ISSUE TITLE
Naming Conventions

### ISSUE THREAD

**Issue by alancleary** [legumeinfo/web-components#25]
I think we need to decide what our naming conventions are going to be within this repo and get everything currently in the repo to conform to them. Things that will be affected:

- Class names
- Tag names
- File names
- File structure (as a side effect)

The convention that seems to be loosely in place already is to prefix everything with "lis" and to suffix with what type something is. The case of the name depends on the convention used for the thing being named, i.e. upper camel case for class names, kebab case for tag names, and kebab case for file names. Take the pagination element, for example:

- Class name: `LisPaginationElement`
- Tag name: `<lis-pagination-element>`
- File names: `lis-pagination-element.ts` and `lis-pagination-element.html`

However, not all components in the project currently conform to this. For instance, the gene search element has `LisGeneSearchElement` and `<lis-gene-search-element>` as its class and tag names, respectively, but its file names are missing the `-element` suffix.

It's also worth considering what's standard/convention elsewhere:

Standard: None of the standard HTML tags (`<table>`, `<form>`, `<p>`, etc) include a `-element` suffix because a tag is always an element, so the `-element` suffix would be gratuitous. However, every JavaScript class for these elements in the [HTML DOM API](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API) has the prefix `HTML` and the suffix `Element` (`HTMLTableElement`, `HTMLFormElement`, `HTMLParagraphElement`, etc). For this reason, it might be worth dropping the `-element` suffix from our tags but keeping it everywhere else. Additionally, I think we should keep the `lis-` prefix in a tag names and keep the `Lis` prefix in place of the `HTML` prefix in our element classes. Both will serve as a sort of namespace for our elements. I'm not sure whether or not it makes sense to keep the `lis-` prefix in file names. I'm leaning towards no since we'll be bundling the generated .js files. And even if someone uses the files as a typescript library, the import paths will always start with `legumeinfo/web-components/`, so having subpaths that start with `lis-` would be redundant.

Convention: Angular has implemented its own interpretation of Web Components since long before they were supported by the web standard. Angular is an opinionated framework and has conventions for pretty much everything. When it comes to naming, they basically use the convention I just described but with the word "component" in place of "element". Their filenames, however, use a `.` instead of a `-` for the type part. For example:

- A component: `something-awesome.component.ts`
- A component's template: `something-awesome.component.html`
- A component's SCSS: `something-awesome.component.scss`
- A directive: `something-useful.directive.ts`
- A module: `the-umbrella.module.ts`
- A service: `get-data.service.ts`

Popular Angular libraries follow this convention as well for their own types, such as NgRX/Store:

- For actions: `something-awesome.actions.ts`
- A reducer: `something-awesome.reducer.ts`
- For effects: `something-awesome.effects.ts`
- For states: `app.states.ts`

Since GCV uses Angular, I've used this convention for my own types of things as well:

- A model: `organism.model.ts`
- A utility: `internet.util.ts`
- A mixin: `extra-functionality.mixin.ts`

The number of examples I've just given may seem excessive but I think it highlights another benefit of this convention - it suggests how things should be organized in directories. Anything that matches `*.component.*` should be in a `components/` directory; anything that matches an NgRX/Store type should be in a `store/` directory, and possibly a type-specific subdirectory if there's many files; and my own types get stored in their own type-specific directories. Not only does this help organize things, it also allows a barrel to be made for each directory, allowing import statements to be as vague or specific as the developer would like.

So, considering all of this, I'll give my preference by revisiting the pagination example:

- Class name: `LisPaginationElement`
- Tag name: `<lis-pagination>`
- File names: `pagination.element.ts` and `pagination-element.html`

Please, someone (anyone! ), share your opinion. It would be nice if we settled on conventions that everyone will enjoy working with rather than me imposing my opinion on everyone.

P.S. Is it worth adding Steven, Sudhansu, and/or Rex to the conversation since they'll be using the tags and elements?

**Comment by adf-ncgr** [comment]
I have to admit to some frustration at the number of different forms that the same basic idea takes in different contexts, but I think if we have a clearly defined convention ruling the morpho-diversity, I will probably get used to it (assuming I spend enough time actually coding). I would probably personally prefer to keep lis as part of the file names, just for minimizing gratuitous generation of additional morphotypes outside of what externally evolved conventions dictate, but I am not going to go all vim-vs-emacs about it.

**Comment by alancleary** [comment]
I actually sympathize with your sentiment @adf-ncgr. While the preference I gave at the end of my OP is consistent with the standards/conventions I discussed, it's pretty inconsistent with itself, and I think consistency and simplicity does have value (and cost when omitted). So, for our own sanity, maybe we should go "full verbose" on everything:

- Class name: `LisPaginationElement`
- Tag name: `<lis-pagination-element>`
- File names: `lis-pagination-element.ts` and `lis-pagination-element.html`

In other words, all names will include the lis prefix and the type suffix in all contexts. And all names will be kebab case, except the class name, which will be upper camel case because you can't use kebab case for a class name. The dual casing convention is consistent with pretty much every web component implementation: web standard, Lit, Angular, etc. Thoughts?

**Comment by adf-ncgr** [comment]
works for me, but we can wait to see if anyone else wants to weigh in; in any case, thanks for being considerate to my rapidly ossifying synapses!

**Comment by alancleary** [comment]
Adding this issue to the documentation Epic since it would be nice to have this in place before turning the key on the docs.

**Comment by alancleary** [comment]
Commit 957e10c5f9db17eaf9b3d4575e892a132de1abe7 updated the necessary existing files and code to make the repo conform to the agreed upon (thanks @adf-ncgr!) naming conventions. Closing.
